pysph/sph/wc/crksph.py

```diff
@@ -61,67 +70,80 @@ def loop_all(self, d_idx, d_x, d_y, d_z, d_h, s_x, s_y, s_z, s_h, s_m,
-            grad_ai[gam] = grad_ai[gam] * ai * ai
-
-        if N_NBRS > 0 or condition:
-            d_ai[d_idx] = ai
-            for i in range(d):
-                d_gradai[d * d_idx + i] = grad_ai[i]
-                d_bi[d * d_idx + i] = bi[i]
-                for j in range(d):
-                    d_gradbi[d * d_idx + d * i + j] = grad_bi[i][j]
+        identity(m2inv, d)
+        augmented_matrix(m2, m2inv, d, d, temp_aug_m2)
+
+        # If is_singular > 0 then matrix was singular
+        is_singular = gj_solve(temp_aug_m2, d, d, m2inv)
+
+        if is_singular > 0.0:
+            # Cannot do much if the matrix is singular.  Perhaps later
+            # we can tag such particles to see if the user can do something.
+            pass
         else:
+            mat_vec_mult(m2inv, m1, d, temp_vec)
+
+            # Eq. 12.
+            ai = 1.0/(m0 - dot(temp_vec, m1, d))
+            # Eq. 13.
+            mat_vec_mult(m2inv, m1, d, bi)
+            for gam in range(d):
+                bi[gam] = -bi[gam]
+
+            # Eq. 14. and 15.
+            for gam in range(d):
+                temp1 = grad_m0[gam]
+                for alp in range(d):
+                    temp2 = 0.0
+                    for bet in range(d):
+                        temp1 -= m2inv[d*alp + bet] * (
+                            m1[bet] * grad_m1[d*gam + alp] +
+                            m1[alp] * grad_m1[d*gam + bet]
+                        )
+                        temp2 -= (
+                            m2inv[d*alp + bet] * grad_m1[d*gam + bet]
+                        )
+                        for phi in range(d):
+                            for psi in range(d):
+                                temp1 += (
+                                    m2inv[d*alp + phi] * m2inv[d*psi + bet] *
+                                    grad_m2[d2*gam + d*phi + psi] *
+                                    m1[bet] * m1[alp]
+                                )
+                                temp2 += (
+                                    m2inv[d*alp + phi] * m2inv[d*psi + bet] *
+                                    grad_m2[d2*gam + d*phi + psi] * m1[bet]
+                                )
+                    grad_bi[d*gam + alp] = temp2
+                grad_ai[gam] = -ai*ai*temp1
+
+        if N_NBRS < 2 or is_singular > 0.0:
             d_ai[d_idx] = 1.0
             for i in range(d):
                 d_gradai[d * d_idx + i] = 0.0
                 d_bi[d * d_idx + i] = 0.0
                 for j in range(d):
-                    d_gradbi[d * d_idx + d * i + j] = 0.0
+                    d_gradbi[d2 * d_idx + d * i + j] = 0.0
+        else:
+            d_ai[d_idx] = ai
+            for i in range(d):
+                d_gradai[d * d_idx + i] = grad_ai[i]
+                d_bi[d * d_idx + i] = bi[i]
+                for j in range(d):
+                    d_gradbi[d2 * d_idx + d * i + j] = grad_bi[d*i + j]
 
 
 class CRKSPH(Equation):
```

Comment:
In line 182 the equation should be
```latex
            \partial_{\gamma}B_{i}^{\alpha} = -\left(m_{2}^{-1}\right)^
            {\alpha \beta}\partial_{\gamma}m_{1}^{\beta} + \left(m_{2}^{-1}\right)^
            {\alpha \phi}\partial_{\gamma}m_{2}^{\phi \psi}\left(m_{2}^
            {-1}\right)^{\psi \beta}m_{1}^{\beta}
```
in 169
```latex
            m_1^{\beta}m_1^{\alpha}\right]^{-1}
```
in 174
```latex
            \partial_{\gamma}A_{i} = -A_{i}^{2}\left(\partial_{\gamma}
            m_{0}-\left(m_{2}^{-1}\right)^{\alpha \beta}\left(
            m_{1}^{\beta}\partial_{\gamma}m_{1}^{\alpha} +
            \partial_{\gamma}m_{1}^{\beta}m_{1}^{\alpha}\right) +
            \left(m_{2}^{-1}\right)^{\alpha \phi}\partial_{\gamma}
            m_{2}^{\phi \psi}\left(m_{2}^{-1}\right)^{\psi \beta}
            m_{1}^{\beta}m_{1}^{\alpha} \right)
```

Reply:
Thanks for the attention to detail.